javax.lang.model.element.UnknownElementException: Unknown element: “repository”

Record dependency injection using the java 15 preview feature with micronaut using java 15 jdk causing the error

```
public record ProductListener(GenericRepository repository) {
    public Flowable<List<Product>> findByFreeText(String text) {
        var item = this.repository.findAll();
    }
}


@Repository
public interface GenericRepository extends RxJavaCrudRepository<Product, Long> {
}
```

Build.gradle

```
tasks.withType(JavaCompile).each {
    it.options.compilerArgs.add('--enable-preview')
}
```

Exception

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileJava'.
    at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:200)
    at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
    at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:198)
    at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:179)
    at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
    at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
    at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
    at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
    at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
    at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
    at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
    at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
    at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
    at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
    at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
    at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
    at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
    at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
    at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
    at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
    at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
    at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
    at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
    at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
    at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
    at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
    at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
    at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
    at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
    at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
    at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
    at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
    at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
    at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
    at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
    at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.RuntimeException: javax.lang.model.element.UnknownElementException: Unknown element: "repository"
    at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:163)
    at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
    at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
    at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:74)
    at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
    at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
    at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:55)
    at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:40)
    at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
    at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
    at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
    at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
    at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
    at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
    at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
    at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
    at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
    at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
    at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
    at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
    at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
    at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
    at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
    at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
    at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
    at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: javax.lang.model.element.UnknownElementException: Unknown element: "repository"
    at io.micronaut.annotation.processing.BeanDefinitionInjectProcessor$AnnBeanElementVisitor.visitUnknown(BeanDefinitionInjectProcessor.java:2002)
    at jdk.compiler/com.sun.tools.javac.code.Symbol$RecordComponent.accept(Symbol.java:1817)
    at io.micronaut.annotation.processing.BeanDefinitionInjectProcessor$AnnBeanElementVisitor.visitType(BeanDefinitionInjectProcessor.java:544)
    at jdk.compiler/com.sun.tools.javac.code.Symbol$ClassSymbol.accept(Symbol.java:1560)
    at io.micronaut.annotation.processing.BeanDefinitionInjectProcessor.lambda$process$7(BeanDefinitionInjectProcessor.java:181)
    at io.micronaut.annotation.processing.BeanDefinitionInjectProcessor.process(BeanDefinitionInjectProcessor.java:176)
    at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
    at org.gradle.api.internal.tasks.compile.processing.DynamicProcessor.process(DynamicProcessor.java:52)
    at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
    at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.access$401(TimeTrackingProcessor.java:37)
    at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:99)
    at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:96)
    at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.track(TimeTrackingProcessor.java:117)
    at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.process(TimeTrackingProcessor.java:96)
    at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1025)
    at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:940)
    at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1269)
    at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1383)
    at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1261)
    at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:935)
    at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
    at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)
```

![image](https://user-images.githubusercontent.com/1625641/96339697-11125700-10e2-11eb-8485-b06984e743f7.png)

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
@@ -385,6 +385,9 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
         @Override
         public Object visitType(TypeElement classElement, Object o) {
             Name classElementQualifiedName = classElement.getQualifiedName();
+            if ("java.lang.Record".equals(classElementQualifiedName.toString())) {
+                return o;
+            }
             if (visitedTypes.contains(classElementQualifiedName)) {
                 // bail out if already visited
                 return o;
@@ -1909,8 +1912,11 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
 
         @Override
         public Object visitUnknown(Element e, Object o) {
-            note("Visit unknown %s for %s", e.getSimpleName(), o);
-            return super.visitUnknown(e, o);
+            if (!JavaModelUtils.isRecordOrRecordComponent(e)) {
+                note("Visit unknown %s for %s", e.getSimpleName(), o);
+                return super.visitUnknown(e, o);
+            }
+            return 0;
         }
 
         /**

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
@@ -1916,7 +1916,7 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
                 note("Visit unknown %s for %s", e.getSimpleName(), o);
                 return super.visitUnknown(e, o);
             }
-            return 0;
+            return o;
         }
 
         /**

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/SuperclassAwareTypeVisitor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/SuperclassAwareTypeVisitor.java
@@ -25,7 +25,6 @@ import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.*;
 import javax.lang.model.util.AbstractTypeVisitor8;
 import javax.lang.model.util.Types;
-import java.lang.reflect.AnnotatedType;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/JavaClassElement.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/JavaClassElement.java
@@ -15,9 +15,11 @@
  */
 package io.micronaut.annotation.processing.visitor;
 
+import edu.umd.cs.findbugs.annotations.NonNull;
 import io.micronaut.annotation.processing.AnnotationUtils;
 import io.micronaut.annotation.processing.ModelUtils;
 import io.micronaut.annotation.processing.PublicMethodVisitor;
+import io.micronaut.annotation.processing.SuperclassAwareTypeVisitor;
 import io.micronaut.core.annotation.AnnotationMetadata;
 import io.micronaut.core.annotation.Internal;
 import io.micronaut.core.naming.NameUtils;
@@ -26,9 +28,8 @@ import io.micronaut.core.util.StringUtils;
 import io.micronaut.inject.ast.*;
 import io.micronaut.inject.processing.JavaModelUtils;
 
-import edu.umd.cs.findbugs.annotations.NonNull;
-import javax.lang.model.element.*;
 import javax.lang.model.element.Element;
+import javax.lang.model.element.*;
 import javax.lang.model.type.DeclaredType;
 import javax.lang.model.type.TypeMirror;
 import javax.lang.model.type.TypeVariable;
@@ -179,102 +180,148 @@ public class JavaClassElement extends AbstractJavaElement implements ArrayableCl
     public List<PropertyElement> getBeanProperties() {
         Map<String, BeanPropertyData> props = new LinkedHashMap<>();
         Map<String, VariableElement> fields = new LinkedHashMap<>();
-
-        classElement.asType().accept(new PublicMethodVisitor<Object, Object>(visitorContext) {
-
-            @Override
-            protected boolean isAcceptable(javax.lang.model.element.Element element) {
-                if (element.getKind() == ElementKind.FIELD) {
-                    return true;
+        if (isRecord()) {
+            classElement.asType().accept(new SuperclassAwareTypeVisitor<Object, Object>(visitorContext) {
+                @Override
+                protected boolean isAcceptable(Element element) {
+                    return JavaModelUtils.isRecord(element);
                 }
-                if (element.getKind() == ElementKind.METHOD && element instanceof ExecutableElement) {
-                    Set<Modifier> modifiers = element.getModifiers();
-                    if (modifiers.contains(Modifier.PUBLIC) && !modifiers.contains(Modifier.STATIC)) {
-                        ExecutableElement executableElement = (ExecutableElement) element;
-                        String methodName = executableElement.getSimpleName().toString();
-                        if (methodName.contains("$")) {
-                            return false;
-                        }
 
-                        if (NameUtils.isGetterName(methodName) && executableElement.getParameters().isEmpty()) {
-                            return true;
-                        } else {
-                            return NameUtils.isSetterName(methodName) && executableElement.getParameters().size() == 1;
+                @Override
+                public Object visitDeclared(DeclaredType type, Object o) {
+                    Element element = type.asElement();
+                    if (isAcceptable(element)) {
+                        List<? extends Element> enclosedElements = element.getEnclosedElements();
+                        for (Element enclosedElement : enclosedElements) {
+                            if (JavaModelUtils.isRecordComponent(enclosedElement) || enclosedElement instanceof ExecutableElement) {
+                                if (enclosedElement.getKind() != ElementKind.CONSTRUCTOR) {
+                                    accept(type, enclosedElement, o);
+                                }
+                            }
                         }
                     }
-                }
-                return false;
-            }
-
-            @Override
-            protected void accept(DeclaredType declaringType, javax.lang.model.element.Element element, Object o) {
-
-                if (element instanceof VariableElement) {
-                    fields.put(element.getSimpleName().toString(), (VariableElement) element);
-                    return;
+                    return o;
                 }
 
+                @Override
+                protected void accept(DeclaredType type, Element element, Object o) {
+                    String name = element.getSimpleName().toString();
+                    if (element instanceof ExecutableElement) {
+                        BeanPropertyData beanPropertyData = props.get(name);
+                        if (beanPropertyData != null) {
+                            beanPropertyData.getter = (ExecutableElement) element;
+                        }
+                    } else {
 
-                ExecutableElement executableElement = (ExecutableElement) element;
-                String methodName = executableElement.getSimpleName().toString();
-                final TypeElement declaringTypeElement = (TypeElement) executableElement.getEnclosingElement();
+                        props.computeIfAbsent(name, propertyName -> {
 
-                if (NameUtils.isGetterName(methodName) && executableElement.getParameters().isEmpty()) {
-                    String propertyName = NameUtils.getPropertyNameForGetter(methodName);
-                    TypeMirror returnType = executableElement.getReturnType();
-                    ClassElement getterReturnType;
-                    if (returnType instanceof TypeVariable) {
-                        TypeVariable tv = (TypeVariable) returnType;
-                        final String tvn = tv.toString();
-                        final ClassElement classElement = getTypeArguments().get(tvn);
-                        if (classElement != null) {
-                            getterReturnType = classElement;
+                            BeanPropertyData beanPropertyData = new BeanPropertyData(propertyName);
+                            beanPropertyData.declaringType = JavaClassElement.this;
+                            beanPropertyData.type = mirrorToClassElement(element.asType(), visitorContext, genericTypeInfo);
+                            return beanPropertyData;
+                        });
+                    }
+                }
+
+            }, null);
+        } else {
+
+            classElement.asType().accept(new PublicMethodVisitor<Object, Object>(visitorContext) {
+
+                @Override
+                protected boolean isAcceptable(javax.lang.model.element.Element element) {
+                    if (element.getKind() == ElementKind.FIELD) {
+                        return true;
+                    }
+                    if (element.getKind() == ElementKind.METHOD && element instanceof ExecutableElement) {
+                        Set<Modifier> modifiers = element.getModifiers();
+                        if (modifiers.contains(Modifier.PUBLIC) && !modifiers.contains(Modifier.STATIC)) {
+                            ExecutableElement executableElement = (ExecutableElement) element;
+                            String methodName = executableElement.getSimpleName().toString();
+                            if (methodName.contains("$")) {
+                                return false;
+                            }
+
+                            if (NameUtils.isGetterName(methodName) && executableElement.getParameters().isEmpty()) {
+                                return true;
+                            } else {
+                                return NameUtils.isSetterName(methodName) && executableElement.getParameters().size() == 1;
+                            }
+                        }
+                    }
+                    return false;
+                }
+
+                @Override
+                protected void accept(DeclaredType declaringType, javax.lang.model.element.Element element, Object o) {
+
+                    if (element instanceof VariableElement) {
+                        fields.put(element.getSimpleName().toString(), (VariableElement) element);
+                        return;
+                    }
+
+
+                    ExecutableElement executableElement = (ExecutableElement) element;
+                    String methodName = executableElement.getSimpleName().toString();
+                    final TypeElement declaringTypeElement = (TypeElement) executableElement.getEnclosingElement();
+
+                    if (NameUtils.isGetterName(methodName) && executableElement.getParameters().isEmpty()) {
+                        String propertyName = NameUtils.getPropertyNameForGetter(methodName);
+                        TypeMirror returnType = executableElement.getReturnType();
+                        ClassElement getterReturnType;
+                        if (returnType instanceof TypeVariable) {
+                            TypeVariable tv = (TypeVariable) returnType;
+                            final String tvn = tv.toString();
+                            final ClassElement classElement = getTypeArguments().get(tvn);
+                            if (classElement != null) {
+                                getterReturnType = classElement;
+                            } else {
+                                getterReturnType = mirrorToClassElement(returnType, visitorContext, JavaClassElement.this.genericTypeInfo);
+                            }
                         } else {
                             getterReturnType = mirrorToClassElement(returnType, visitorContext, JavaClassElement.this.genericTypeInfo);
                         }
-                    } else {
-                        getterReturnType = mirrorToClassElement(returnType, visitorContext, JavaClassElement.this.genericTypeInfo);
-                    }
 
-                    BeanPropertyData beanPropertyData = props.computeIfAbsent(propertyName, BeanPropertyData::new);
-                    configureDeclaringType(declaringTypeElement, beanPropertyData);
-                    beanPropertyData.type = getterReturnType;
-                    beanPropertyData.getter = executableElement;
-                    if (beanPropertyData.setter != null) {
-                        TypeMirror typeMirror = beanPropertyData.setter.getParameters().get(0).asType();
-                        ClassElement setterParameterType = mirrorToClassElement(typeMirror, visitorContext, JavaClassElement.this.genericTypeInfo);
-                        if (!setterParameterType.getName().equals(getterReturnType.getName())) {
-                            beanPropertyData.setter = null; // not a compatible setter
+                        BeanPropertyData beanPropertyData = props.computeIfAbsent(propertyName, BeanPropertyData::new);
+                        configureDeclaringType(declaringTypeElement, beanPropertyData);
+                        beanPropertyData.type = getterReturnType;
+                        beanPropertyData.getter = executableElement;
+                        if (beanPropertyData.setter != null) {
+                            TypeMirror typeMirror = beanPropertyData.setter.getParameters().get(0).asType();
+                            ClassElement setterParameterType = mirrorToClassElement(typeMirror, visitorContext, JavaClassElement.this.genericTypeInfo);
+                            if (!setterParameterType.getName().equals(getterReturnType.getName())) {
+                                beanPropertyData.setter = null; // not a compatible setter
+                            }
                         }
-                    }
-                } else if (NameUtils.isSetterName(methodName) && executableElement.getParameters().size() == 1) {
-                    String propertyName = NameUtils.getPropertyNameForSetter(methodName);
-                    TypeMirror typeMirror = executableElement.getParameters().get(0).asType();
-                    ClassElement setterParameterType = mirrorToClassElement(typeMirror, visitorContext, JavaClassElement.this.genericTypeInfo);
+                    } else if (NameUtils.isSetterName(methodName) && executableElement.getParameters().size() == 1) {
+                        String propertyName = NameUtils.getPropertyNameForSetter(methodName);
+                        TypeMirror typeMirror = executableElement.getParameters().get(0).asType();
+                        ClassElement setterParameterType = mirrorToClassElement(typeMirror, visitorContext, JavaClassElement.this.genericTypeInfo);
 
-                    BeanPropertyData beanPropertyData = props.computeIfAbsent(propertyName, BeanPropertyData::new);
-                    configureDeclaringType(declaringTypeElement, beanPropertyData);
-                    ClassElement propertyType = beanPropertyData.type;
-                    if (propertyType != null) {
-                        if (propertyType.getName().equals(setterParameterType.getName())) {
+                        BeanPropertyData beanPropertyData = props.computeIfAbsent(propertyName, BeanPropertyData::new);
+                        configureDeclaringType(declaringTypeElement, beanPropertyData);
+                        ClassElement propertyType = beanPropertyData.type;
+                        if (propertyType != null) {
+                            if (propertyType.getName().equals(setterParameterType.getName())) {
+                                beanPropertyData.setter = executableElement;
+                            }
+                        } else {
                             beanPropertyData.setter = executableElement;
                         }
-                    } else {
-                        beanPropertyData.setter = executableElement;
                     }
                 }
-            }
 
-            private void configureDeclaringType(TypeElement declaringTypeElement, BeanPropertyData beanPropertyData) {
-                if (beanPropertyData.declaringType == null && !classElement.equals(declaringTypeElement)) {
-                    beanPropertyData.declaringType = mirrorToClassElement(
-                            declaringTypeElement.asType(),
-                            visitorContext,
-                            genericTypeInfo
-                    );
+                private void configureDeclaringType(TypeElement declaringTypeElement, BeanPropertyData beanPropertyData) {
+                    if (beanPropertyData.declaringType == null && !classElement.equals(declaringTypeElement)) {
+                        beanPropertyData.declaringType = mirrorToClassElement(
+                                declaringTypeElement.asType(),
+                                visitorContext,
+                                genericTypeInfo
+                        );
+                    }
                 }
-            }
-        }, null);
+            }, null);
+        }
 
         if (!props.isEmpty()) {
             List<PropertyElement> propertyElements = new ArrayList<>();
@@ -346,7 +393,6 @@ public class JavaClassElement extends AbstractJavaElement implements ArrayableCl
         } else {
             return Collections.emptyList();
         }
-
     }
 
     @Override

--- a/inject-java/src/test/groovy/io/micronaut/inject/AbstractTypeElementSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/AbstractTypeElementSpec.groovy
@@ -23,15 +23,19 @@ import groovy.transform.CompileStatic
 import io.micronaut.annotation.processing.AnnotationUtils
 import io.micronaut.annotation.processing.GenericUtils
 import io.micronaut.annotation.processing.ModelUtils
+import io.micronaut.annotation.processing.visitor.JavaClassElement
+import io.micronaut.annotation.processing.visitor.JavaVisitorContext
 import io.micronaut.context.ApplicationContext
 import io.micronaut.context.DefaultApplicationContext
 import io.micronaut.core.annotation.AnnotationMetadata
 import io.micronaut.core.beans.BeanIntrospection
+import io.micronaut.core.convert.value.MutableConvertibleValuesMap
 import io.micronaut.core.io.scan.ClassPathResourceLoader
 import io.micronaut.core.naming.NameUtils
 import io.micronaut.inject.annotation.AnnotationMapper
 import io.micronaut.inject.annotation.AnnotationMetadataWriter
 import io.micronaut.annotation.processing.JavaAnnotationMetadataBuilder
+import io.micronaut.inject.ast.ClassElement
 import io.micronaut.inject.writer.BeanConfigurationWriter
 import io.micronaut.support.Parser
 import spock.lang.Specification
@@ -81,6 +85,43 @@ abstract class AbstractTypeElementSpec extends Specification {
         JavaAnnotationMetadataBuilder builder = newJavaAnnotationBuilder()
         AnnotationMetadata metadata = argument != null ? builder.build(argument) : null
         return metadata
+    }
+
+    protected ClassElement buildClassElement(String cls) {
+        List<Element> elements = Parser.parseLines("",
+                cls
+        ).toList()
+        def context = new Context()
+        def env = JavacProcessingEnvironment.instance(context)
+        try {
+            final com.sun.tools.javac.main.JavaCompiler jc = com.sun.tools.javac.main.JavaCompiler.instance(context)
+            jc?.initModules(com.sun.tools.javac.util.List.nil())
+        } catch (e) {
+            // ignore, must be JDK 8
+        }
+        def typeElements = JavacElements.instance(context)
+        ModelUtils modelUtils = new ModelUtils(typeElements, env.typeUtils) {}
+        GenericUtils genericUtils = new GenericUtils(typeElements, env.typeUtils, modelUtils) {}
+        AnnotationUtils annotationUtils = new AnnotationUtils(env, typeElements, env.messager, env.typeUtils, modelUtils, genericUtils, env.filer) {
+        }
+        def element = elements ? elements[0] : null
+        assert element instanceof TypeElement
+        def builder = newJavaAnnotationBuilder()
+        return new JavaClassElement(
+                element,
+                builder.build(element),
+                new JavaVisitorContext(
+                        env,
+                        env.messager,
+                        typeElements,
+                        annotationUtils,
+                        env.typeUtils,
+                        modelUtils,
+                        genericUtils,
+                        env.filer,
+                        new MutableConvertibleValuesMap<Object>()
+                )
+        )
     }
 
     protected TypeElement buildTypeElement(String cls) {

--- a/inject-java/src/test/groovy/io/micronaut/inject/records/RecordBeansSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/records/RecordBeansSpec.groovy
@@ -1,0 +1,55 @@
+package io.micronaut.inject.records
+
+import io.micronaut.inject.AbstractTypeElementSpec
+import io.micronaut.inject.BeanDefinition
+import spock.lang.IgnoreIf
+
+@IgnoreIf({ !jvm.isJava14Compatible() })
+class RecordBeansSpec extends AbstractTypeElementSpec {
+
+    void 'test bean that is a record'() {
+        given:
+        BeanDefinition<?> definition = buildBeanDefinition('test.Test', '''
+package test;
+
+@javax.inject.Singleton
+record Test(OtherBean otherBean) {
+
+}
+
+@javax.inject.Singleton
+class OtherBean {
+
+}
+''')
+
+        expect:
+        definition.constructor.arguments.length == 1
+    }
+
+    void 'test bean factory that returns a record'() {
+        given:
+        BeanDefinition<?> definition = buildBeanDefinition('test.TestFactory', '''
+package test;
+
+@io.micronaut.context.annotation.Factory
+class TestFactory {
+    @io.micronaut.context.annotation.Bean
+    Test test(OtherBean otherBean) {
+        return new Test(otherBean);
+    }
+}
+record Test(OtherBean otherBean) {
+
+}
+
+@javax.inject.Singleton
+class OtherBean {
+
+}
+''')
+
+        expect:
+        definition.constructor.arguments.length == 0
+    }
+}

--- a/inject-java/src/test/groovy/io/micronaut/support/Parser.java
+++ b/inject-java/src/test/groovy/io/micronaut/support/Parser.java
@@ -36,6 +36,8 @@ import com.sun.source.util.Trees;
 import com.sun.tools.javac.api.JavacTool;
 import com.sun.tools.javac.util.Context;
 import io.micronaut.annotation.processing.*;
+import io.micronaut.core.util.CollectionUtils;
+import spock.util.environment.Jvm;
 
 import java.io.File;
 import java.io.IOException;
@@ -72,6 +74,7 @@ public final class Parser {
         InMemoryJavaFileManager fileManager =
                 new InMemoryJavaFileManager(
                         compiler.getStandardFileManager(diagnosticCollector, Locale.getDefault(), UTF_8));
+        Set<String> options = getCompilerOptions();
         Context context = new Context();
         JavacTask task =
                 ((JavacTool) compiler)
@@ -79,7 +82,7 @@ public final class Parser {
                                 null, // explicitly use the default because old javac logs some output on stderr
                                 fileManager,
                                 diagnosticCollector,
-                                ImmutableSet.of(),
+                                options,
                                 ImmutableSet.of(),
                                 Arrays.asList(sources),
                                 context);
@@ -97,6 +100,20 @@ public final class Parser {
         }
     }
 
+    private static Set<String> getCompilerOptions() {
+        Set<String> options;
+        if (Jvm.getCurrent().isJava15Compatible()) {
+            options = CollectionUtils.setOf(
+                    "--enable-preview",
+                    "-source",
+                    "15"
+            );
+        } else {
+            options = Collections.emptySet();
+        }
+        return options;
+    }
+
     /**
      * Parses {@code sources} into {@linkplain CompilationUnitTree compilation units}. This method
      * <b>does not</b> compile the sources.
@@ -109,13 +126,14 @@ public final class Parser {
                 new InMemoryJavaFileManager(
                         standardFileManager);
         Context context = new Context();
+        Set<String> options = getCompilerOptions();
         JavacTask task =
                 ((JavacTool) compiler)
                         .getTask(
                                 null, // explicitly use the default because old javac logs some output on stderr
                                 fileManager,
                                 diagnosticCollector,
-                                ImmutableSet.of(),
+                                options,
                                 ImmutableSet.of(),
                                 Arrays.asList(sources),
                                 context);

--- a/inject-java/src/test/groovy/io/micronaut/visitors/PropertyElementSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/visitors/PropertyElementSpec.groovy
@@ -17,6 +17,8 @@ package io.micronaut.visitors
 
 import io.micronaut.http.annotation.Get
 import io.micronaut.inject.AbstractTypeElementSpec
+import io.micronaut.inject.ast.ClassElement
+import io.micronaut.inject.ast.PropertyElement
 import spock.lang.IgnoreIf
 import spock.lang.Specification
 import spock.util.environment.Jvm
@@ -25,6 +27,26 @@ import javax.annotation.Nullable
 import javax.validation.constraints.NotBlank
 
 class PropertyElementSpec extends AbstractTypeElementSpec {
+    @IgnoreIf({ !jvm.isJava14Compatible() })
+    void 'test bean properties work for records'() {
+        given:
+        ClassElement classElement = buildClassElement('''
+package test;
+
+record Book( @javax.validation.constraints.NotBlank String title, int pages) {}
+''')
+        def beanProperties = classElement.getBeanProperties()
+        def titleProp = beanProperties.find { it.name == 'title' }
+        expect:
+        classElement.isRecord()
+        beanProperties.size() == 2
+        titleProp != null
+        titleProp.type.name == String.name
+        titleProp.hasAnnotation(NotBlank)
+        beanProperties.every { it.readOnly }
+    }
+
+
     // Java 9+ doesn't allow resolving elements was the compiler
     // is finished being used so this test cannot be made to work beyond Java 8 the way it is currently written
     @IgnoreIf({ Jvm.current.isJava9Compatible() })

--- a/inject/src/main/java/io/micronaut/inject/processing/JavaModelUtils.java
+++ b/inject/src/main/java/io/micronaut/inject/processing/JavaModelUtils.java
@@ -180,6 +180,15 @@ public class JavaModelUtils {
      * @return True if it is
      */
     public static boolean isRecordOrRecordComponent(Element e) {
-        return isRecord(e) || resolveKind(e, RECORD_COMPONENT_KIND).isPresent();
+        return isRecord(e) || isRecordComponent(e);
+    }
+
+    /**
+     * Return whether this is a component of a record.
+     * @param e The element
+     * @return True if it is
+     */
+    public static boolean isRecordComponent(Element e) {
+        return resolveKind(e, RECORD_COMPONENT_KIND).isPresent();
     }
 }

--- a/inject/src/main/java/io/micronaut/inject/processing/JavaModelUtils.java
+++ b/inject/src/main/java/io/micronaut/inject/processing/JavaModelUtils.java
@@ -35,6 +35,13 @@ public class JavaModelUtils {
             return Optional.empty();
         }
     }).orElse(null);
+    private static final ElementKind RECORD_COMPONENT_KIND = ReflectionUtils.findDeclaredField(ElementKind.class, "RECORD_COMPONENT").flatMap(field -> {
+        try {
+            return Optional.of((ElementKind) field.get(ElementKind.class));
+        } catch (IllegalAccessException e) {
+            return Optional.empty();
+        }
+    }).orElse(null);
 
     /**
      * The Java APT throws an internal exception {code com.sun.tools.javac.code.Symbol$CompletionFailure} if a class is missing from the classpath and {@link Element#getKind()} is called. This method
@@ -165,5 +172,14 @@ public class JavaModelUtils {
      */
     public static String getClassArrayName(TypeElement typeElement) {
         return "[L" + getClassName(typeElement) + ";";
+    }
+
+    /**
+     * Return whether this is a record or a component of a record.
+     * @param e The element
+     * @return True if it is
+     */
+    public static boolean isRecordOrRecordComponent(Element e) {
+        return isRecord(e) || resolveKind(e, RECORD_COMPONENT_KIND).isPresent();
     }
 }

